zebra/zebra_ns_notify.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Zebra NS collector and notifier for Network NameSpaces
+ * Copyright (C) 2017 6WIND
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#ifdef HAVE_NETLINK
+#ifdef HAVE_NETNS
+#undef _GNU_SOURCE
+#define _GNU_SOURCE
+
+#include <sched.h>
+#endif
+#include <dirent.h>
+#include <sys/inotify.h>
+#include <sys/stat.h>
+
+#include "thread.h"
+#include "ns.h"
+#include "command.h"
+#include "memory.h"
+
+#include "zserv.h"
+#include "zebra_memory.h"
+#include "zebra_ns_notify.h"
+#include "zebra_ns_id.h"
+
+/* upon creation of folder under /var/run/netns,
+ * wait that netns context is bound to
+ * that folder 10 seconds
+ */
+#define ZEBRA_NS_POLLING_INTERVAL_MSEC     1000
+#define ZEBRA_NS_POLLING_MAX_RETRIES  200
+
+extern struct zebra_privs_t zserv_privs;
+
+DEFINE_MTYPE_STATIC(ZEBRA, NETNS_MISC, "ZebraNetNSInfo")
+
+struct zebra_netns_info {
+	const char *netnspath;
+	unsigned int retries;
+};
+
+static int zebra_ns_ready_read(struct thread *t);
+static void zebra_ns_notify_create_context_from_entry_name(const char *name);
+static int zebra_ns_continue_read(struct zebra_netns_info *zns_info,
+				  int stop_retry);
+static int zebra_ns_notify_read(struct thread *t);
+
+static void zebra_ns_notify_create_context_from_entry_name(const char *name)
+{
+	char *netnspath = ns_netns_pathname(NULL, name);
+	struct vrf *vrf;
+	int ret;
+	ns_id_t ns_id;
+
+	if (netnspath == NULL)
+		return;
+
+	if (vrf_handler_create(NULL, name, &vrf) != CMD_SUCCESS) {
+		zlog_warn("NS notify : failed to create VRF %s", name);
+		return;
+	}
+	if (zserv_privs.change(ZPRIVS_RAISE))
+		zlog_err("Can't raise privileges");
+        ns_id = zebra_ns_id_get(netnspath);
+	ret = ns_handler_create(NULL, vrf, netnspath, ns_id);
```

Comment:
In the `"netns NAME"` CLI command you don't raise privileges before calling `ns_handler_create()`. There's something wrong here. Please address this problem.

Reply:
the NSID generation requires higher privileges.
I braced ns_id_get() accordingly, then.

I have a doubt, do you know if setns() needs higher privileges ?
If yes, then ns_handler_create() will need higher privileges, then. but this can be done later in the code of ns.c